ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/parser.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright © Aptos Foundation
+
+use crate::{db::upsert_uris, models::NFTMetadataCrawlerURIs, schema::nft_metadata_crawler_uris};
+use chrono::Utc;
+use diesel::{
+    r2d2::{ConnectionManager, PooledConnection},
+    PgConnection, QueryDsl, RunQueryDsl,
+};
+use google_cloud_auth::token_source::TokenSource;
+use image::{ImageBuffer, ImageFormat};
+use nft_metadata_crawler_utils::{
+    gcs::{write_image_to_gcs, write_json_to_gcs},
+    NFTMetadataCrawlerEntry,
+};
+use serde_json::Value;
+use std::error::Error;
+
+// Stuct that represents a parser for a single entry from queue
+pub struct Parser<'a> {
+    pub entry: NFTMetadataCrawlerEntry,
+    model: NFTMetadataCrawlerURIs,
+    format: ImageFormat,
+    bucket: String,
+    ts: &'a dyn TokenSource,
+    force: bool,
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(e: NFTMetadataCrawlerEntry, b: String, f: bool, t: &'a dyn TokenSource) -> Self {
+        Self {
+            model: NFTMetadataCrawlerURIs {
+                token_uri: e.token_uri.clone(),
+                raw_image_uri: None,
+                raw_animation_uri: None,
+                cdn_json_uri: None,
+                cdn_image_uri: None,
+                cdn_animation_uri: None,
+                image_resizer_retry_count: 0,
+                json_parser_retry_count: 0,
+                last_updated: Utc::now().naive_utc(),
+            },
+            entry: e,
+            format: ImageFormat::Jpeg,
+            bucket: b,
+            ts: t,
+            force: f,
+        }
+    }
+
+    pub async fn parse(
```

Comment:
This function needs to be split into several other functions. If you want clean code it might be worth creating a separate utils file.